docs/config.md
## Configuring Frontik

Here is the list of all options currently supported by Frontik.
All of them can be set in the configuration file (see [frontik_dev.cfg.ex](/frontik_dev.cfg.ex) for example),
but you can pass them to `/usr/bin/frontik` script through command line as well. The exception is `config` option — it makes sense to pass it only through command line.

These options can be set for each Frontik instance (see [options.py](/frontik/options.py)).

| Option name                  | Type    | Default value | Description                                                            |
| ---------------------------- | ------- | ------------  | ---------------------------------------------------------------------- |
| `host`                       | `str`   | `'0.0.0.0'`   | Host value for incoming connections                                    |
| `port`                       | `int`   | `8000`        | Port to listen to                                                      |
| `reuse_port`                 | `bool`  | `True`        | Binds server socket with SO_REUSEPORT option                           |
| `config`                     | `str`   | `None`        | Path to config file                                                    |
| `app`                        | `str`   | `None`        | Application package name (see [Frontik application structure](/docs/frontik-app.md)) |
| `app_class`                  | `str`   | `None`        | Application class name defined in application root module (by default `FrontikApplication` class is used) |
| `xheaders  `                 | `bool`  | `False`       | Controls Tornado HTTPServer `xheaders` option                          |
| `tornado_settings`           | `dict`  | `None`        | tornado.web.Application settings                                       |
| `autoreload`                 | `bool`  | `False`       | Restart Frontik after changes in application sources or config files   |
| `debug`                      | `bool`  | `False`       | Enable debug mode                                                      |
| `debug_login`                | `str`   | `None`        | Debug mode login for basic authentication (when `debug=False`)         |
| `debug_password`             | `str`   | `None`        | Debug mode password for basic authentication (when `debug=False`)      |
| `handlers_count`             | `int`   | `100`         | Limit for number of simultaneous requests handled by Frontik instance  |
| `datacenter`                 | `str`   | `None`        | Datacenter where current application is running                        |

Logging options:

| Option name                  | Type    | Default value | Description                                                            |
|------------------------------|---------|---------------|------------------------------------------------------------------------|
| `loglevel`                   | `str`   | `info`        | Python log level                                                       |
| `logformat`                  | `str`   | see code      | Log entry format for files and syslog                                  |
| `logfile`                    | `str`   | `None`        | Log file location (set to `None` to disable logging to file)           |
| `stderr_log`                 | `bool`  | `False`       | Send log output to stderr (colorized if possible)                      |
| `stderr_format`              | `str`   | see code      | Log entry format for stderr output                                     |
| `stderr_dateformat`          | `str`   | see code      | Log entry date format for stderr output                                |
| `syslog`                     | `bool`  | `False`       | Enables sending logs to syslog                                         |
| `syslog_address`             | `str`   | `'/dev/log'`  | Syslog address, unix socket name or server address                     |
| `syslog_port`                | `int`   | `None`        | Syslog port. If this value is None, unix socket is used, UDP otherwise |
| `syslog_facility`            | `str`   | `'user'`      | Syslog facility                                                        |
| `suppressed_loggers`         | `list`  | `[]`          | List of logger names to be excluded from debug output                  |
| `sentry_dsn`                 | `str`   | `None`        | Enable Sentry and set Sentry DSN for sending errors                    |
| `statsd_host`                | `str`   | `None`        | Stats server host for metrics                                          |
| `statsd_port`                | `int`   | `None`        | Stats server port for metrics                                          |
| `asyncio_task_threshold_sec` | `int`   | `None`        | Threshold for logging long-running asyncio tasks                       |

HTTP client options:

| Option name                                   | Type    | Default value      | Description                                                                                |
| --------------------------------------------- | ------- | ------------------ | ------------------------------------------------------------------------------------------ |
| `max_http_clients`                            | `int`   | `100`              | Curl max clients option                                                                    |
| `max_http_clients_connects`                   | `int`   | `None`             | Curl max connects option                                                                   |
| `http_client_default_connect_timeout_sec`     | `float` | `0.2`              | Default connect timeout                                                                    |
| `http_client_default_request_timeout_sec`     | `float` | `2.0`              | Default request timeout                                                                    |
| `http_client_default_max_tries`               | `int`   | `2`                | Maximum number of retries per request + 1                                                  |
| `http_client_default_max_timeout_tries`       | `int`   | `1`                | Maximum number of retries due to timeout per request + 1                                   |
| `http_client_default_max_fails`               | `int`   | `0`                | Number of consecutive fails before server is considered dead                               |
| `http_client_default_fail_timeout_sec`        | `float` | `10`               | Timeout before starting making requests to a dead server                                   |
| `http_client_default_retry_policy`            | `str`   | `timeout,http_503` | Conditions when request retry is possible                                                  |
| `http_proxy_host`                             | `str`   | `None`             | HTTP proxy host for Curl HTTP client                                                       |
| `http_proxy_port`                             | `int`   | `3128`             | HTTP proxy port for Curl HTTP client                                                       |
| `http_client_allow_cross_datacenter_requests` | `bool`  | `False`            | Allow requests to different datacenter when no upstream in current datacenter is available |
| `timeout_multiplier`                          | `float` | `1.0`              | Generic timeout multiplier for http requests (useful for testing)                          |

Producers options:

| Option name                         | Type   | Default value | Description                                           |
| `xsl_root`                          | `str`  | `None`        | Root directory for XSL files                          |
| `xml_root`                          | `str`  | `None`        | Root directory for XML files                          |
| `xsl_cache_limit`                   | `int`  | `None`        | Upper limit for XSL LRU files cache                   |
| `xml_cache_limit`                   | `int`  | `None`        | Upper limit for XML LRU files cache                   |
| `xsl_executor_pool_size`            | `int`  | `1`           | Number of background threads for XSLT processing      |
| `jinja_template_root`               | `str`  | `None`        | Root directory for Jinja templates                    |
| `jinja_template_cache_limit`        | `int`  | `50`          | Upper limit for Jinja templates cache                 |
| `jinja_streaming_render_timeout_ms` | `int`  | `50`          | Upper limit (in msecs) for one iteration of partial Jinja template rendering |

The only option that is mandatory for running Frontik is `app` option — the name of application package.

There are also certain options, that can be defined during application initialization, see
[Configuring Frontik application](/docs/config-app.md) for more details.


debian/changelog
frontik (2.31.14) unstable; urgency=medium

  * HH-73599 Send service name to statsd with tag rather then with
    prefix
  * HH-73592 Add non idempotent retry
  * HH-73397 Calculate label color for xml

 -- Automated Build <d.smirnov@hh.ru>  Thu, 19 Oct 2017 17:13:02 +0300

frontik (2.31.13) unstable; urgency=medium

  * HH-73326 Add custom slow start

 -- Automated Build <d.smirnov@hh.ru>  Wed, 11 Oct 2017 14:02:58 +0300

frontik (2.31.12) unstable; urgency=medium

  * HH-73297 Implement new slow start strategies

 -- Automated Build <d.smirnov@hh.ru>  Tue, 10 Oct 2017 16:06:49 +0300

frontik (2.31.11) unstable; urgency=medium

  * HH-73201 do not encode json responses recursively

 -- Automated Build <an.sumin@hh.ru>  Mon, 09 Oct 2017 13:40:36 +0300

frontik (2.31.10) unstable; urgency=medium

  * HH-72947 Fix slow start

 -- Automated Build <d.smirnov@hh.ru>  Fri, 06 Oct 2017 15:31:07 +0300

frontik (2.31.9) unstable; urgency=medium

  * HH-73177 Add requests slow start

 -- Automated Build <d.smirnov@hh.ru>  Thu, 05 Oct 2017 14:39:52 +0300

frontik (2.31.8) unstable; urgency=medium

  * HH-73119 small xml producer optimization

 -- Automated Build <an.sumin@hh.ru>  Wed, 04 Oct 2017 13:22:20 +0300

frontik (2.31.7) unstable; urgency=medium

  * HH-72978 Add server randomization, add check for already finished
    page

 -- Automated Build <d.smirnov@hh.ru>  Fri, 29 Sep 2017 13:10:39 +0300

frontik (2.31.5) unstable; urgency=medium

  * HH-72752 Add connect and request timeout per upstream

 -- Automated Build <d.smirnov@hh.ru>  Tue, 26 Sep 2017 17:44:52 +0300

frontik (2.31.4) unstable; urgency=medium

  * HH-72631 refactor finish_with_postprocessors

 -- Automated Build <an.sumin@hh.ru>  Thu, 21 Sep 2017 12:47:41 +0300

frontik (2.31.3) unstable; urgency=medium

  * HH-72655 Add protobuf client

 -- Automated Build <d.smirnov@hh.ru>  Wed, 20 Sep 2017 17:36:16 +0300

frontik (2.31.2) unstable; urgency=medium

  * HH-72707 Fix requests stats

 -- Automated Build <d.smirnov@hh.ru>  Wed, 20 Sep 2017 13:45:38 +0300

frontik (2.31.1) unstable; urgency=medium

  * HH-72649 Add logging params

 -- Automated Build <d.smirnov@hh.ru>  Mon, 18 Sep 2017 18:39:49 +0300

frontik (2.31.0) unstable; urgency=medium

  * HH-72340 Add http balancing

 -- Automated Build <d.smirnov@hh.ru>  Fri, 15 Sep 2017 20:38:08 +0300

frontik (2.30.3) unstable; urgency=medium

  * HH-72476 prepare tornado4
  * HH-72475 fix preprocessors early exit

 -- Automated Build <an.sumin@hh.ru>  Tue, 12 Sep 2017 10:44:22 +0300

frontik (2.30.2) unstable; urgency=medium

  * HH-72424 fix future return in new preprocessors

 -- Automated Build <an.sumin@hh.ru>  Fri, 08 Sep 2017 15:47:19 +0300

frontik (2.30.1) unstable; urgency=medium

  * HH-66084 add preprocessor decorator

 -- Automated Build <an.sumin@hh.ru>  Thu, 07 Sep 2017 17:22:50 +0300

frontik (2.30.0) unstable; urgency=medium

  * HH-72025 clean things up after removing Tornado patch

 -- Automated Build <an.sumin@hh.ru>  Mon, 04 Sep 2017 17:16:39 +0300

frontik (2.28.24) unstable; urgency=medium

  * HH-70485 remove logging patch

 -- Automated Build <an.sumin@hh.ru>  Wed, 16 Aug 2017 18:17:56 +0300

frontik (2.28.23) unstable; urgency=medium

  * HH-71548 fix tornado dependency
  * HH-71548 build python3 deb package

 -- Automated Build <an.sumin@hh.ru>  Mon, 07 Aug 2017 13:21:33 +0300

frontik (2.28.21) unstable; urgency=medium

  * HH-70830 update js-libs

 -- Automated Build <a.anatskiy@hh.ru>  Thu, 27 Jul 2017 12:03:26 +0300

frontik (2.28.20) unstable; urgency=medium

  * HH-69223 insane debug refactoring

 -- Automated Build <an.sumin@hh.ru>  Mon, 15 May 2017 17:00:19 +0300

frontik (2.28.19) unstable; urgency=medium

  * HH-66959 fix debug output if some data not exist
  * HH-68956 disable breadcrumbs in sentry client

 -- Automated Build <n.kovalev@hh.ru>  Tue, 02 May 2017 16:43:53 +0300

frontik (2.28.18) unstable; urgency=medium

  * HH-68476 use global ContextFilter and request-local log handler

 -- Automated Build <an.sumin@hh.ru>  Fri, 21 Apr 2017 11:54:13 +0300

frontik (2.28.17) unstable; urgency=medium

  * HH-68298 routing grooming

 -- Automated Build <an.sumin@hh.ru>  Thu, 13 Apr 2017 14:43:28 +0300

frontik (2.28.16) unstable; urgency=medium

  * HH-67413 use request_id and handler_name from RequestContext

 -- Automated Build <an.sumin@hh.ru>  Mon, 10 Apr 2017 15:44:18 +0300

frontik (2.28.15) unstable; urgency=low

  * HH-68115 remove supervisor module

 -- Automated Build <an.sumin@hh.ru>  Tue, 04 Apr 2017 11:36:50 +0300

frontik (2.28.14) unstable; urgency=low

  * HH-68068 remove codecov and exception_on_init test

 -- Automated Build <an.sumin@hh.ru>  Mon, 03 Apr 2017 12:35:30 +0300

frontik (2.28.13) unstable; urgency=low

  * HH-67796 fix RequestContext usages

 -- Automated Build <an.sumin@hh.ru>  Tue, 28 Mar 2017 14:51:51 +0300

frontik (2.28.12) unstable; urgency=low

  * HH-67889 cleanup routing

 -- Automated Build <an.sumin@hh.ru>  Tue, 28 Mar 2017 14:16:05 +0300

frontik (2.28.10) unstable; urgency=low

  * HH-67628 remove versions from libcurl debian deps

 -- Automated Build <an.sumin@hh.ru>  Tue, 21 Mar 2017 19:12:47 +0300

frontik (2.28.9) unstable; urgency=low

  * HH-67595 small fix in RequestContext

 -- Automated Build <an.sumin@hh.ru>  Tue, 21 Mar 2017 14:23:55 +0300

frontik (2.28.8) unstable; urgency=low

  * HH-67430 move routing to frontik.routing

 -- Automated Build <an.sumin@hh.ru>  Tue, 14 Mar 2017 09:47:17 +0300

frontik (2.28.7) unstable; urgency=low

  * HH-67233 implement reverse_url method

 -- Automated Build <an.sumin@hh.ru>  Mon, 13 Mar 2017 12:21:17 +0300

frontik (2.28.6) unstable; urgency=low

  * HH-67237 run travis builds on trusty
  * HH-67237 remove deprecated write_error behaviour

 -- Automated Build <an.sumin@hh.ru>  Fri, 10 Mar 2017 14:18:31 +0300

frontik (2.28.5) unstable; urgency=low

  * HH-67114 introduce RequestContext in test mode


frontik/options.py
import tornado.options

tornado.options.define('app', default=None, type=str)
tornado.options.define('app_class', default=None, type=str)
tornado.options.define('tornado_settings', default=None, type=dict)
tornado.options.define('handlers_count', default=100, type=int)
tornado.options.define('reuse_port', default=True, type=bool)
tornado.options.define('xheaders', default=False, type=bool)

tornado.options.define('config', None, str)
tornado.options.define('host', '0.0.0.0', str)
tornado.options.define('port', 8080, int)

tornado.options.define('autoreload', False, bool)
tornado.options.define('stop_timeout', 3, int)
tornado.options.define('slow_callback_threshold_ms', None, int)

tornado.options.define('loglevel', default='info', type=str, help='Log level')
tornado.options.define('logformat', default='[%(process)s] %(asctime)s %(levelname)s %(name)s: %(message)s', type=str,
                       help='Log format for files and syslog')
tornado.options.define('logfile', default=None, type=str, help='Log file name')

tornado.options.define('stderr_log', default=False, type=bool,
                       help='Send log output to stderr (colorized if possible).')
tornado.options.define('stderr_format', default='%(color)s[%(levelname)1.1s %(asctime)s %(name)s '
                                                '%(module)s:%(lineno)d]%(end_color)s %(message)s', type=str)
tornado.options.define('stderr_dateformat', default='%y.%m.%d %H:%M:%S', type=str)

tornado.options.define('syslog', default=False, type=bool)
tornado.options.define('syslog_address', default='/dev/log', type=str)
tornado.options.define('syslog_port', default=None, type=int)
tornado.options.define('syslog_facility', default='user', type=str)

tornado.options.define('suppressed_loggers', default=['tornado.curl_httpclient'], type=list)

tornado.options.define('debug', default=False, type=bool)
tornado.options.define('debug_login', default=None, type=str)
tornado.options.define('debug_password', default=None, type=str)

tornado.options.define('datacenter', default=None, type=str)

tornado.options.define('http_client_default_connect_timeout_sec', default=0.2, type=float)
tornado.options.define('http_client_default_request_timeout_sec', default=2.0, type=float)
tornado.options.define('http_client_default_max_tries', default=2, type=int)
tornado.options.define('http_client_default_max_timeout_tries', default=1, type=int)
tornado.options.define('http_client_default_max_fails', default=0, type=int)
tornado.options.define('http_client_default_fail_timeout_sec', default=10, type=float)
tornado.options.define('http_client_default_retry_policy', default='timeout,http_503', type=str)
tornado.options.define('http_proxy_host', default=None, type=str)
tornado.options.define('http_proxy_port', default=3128, type=int)
tornado.options.define('http_client_allow_cross_datacenter_requests', default=False, type=bool)

tornado.options.define('statsd_host', default=None, type=str)
tornado.options.define('statsd_port', default=None, type=int)
tornado.options.define('statsd_flush_interval_sec', default=30, type=int)
tornado.options.define('generic_metrics_send_interval_ms', default=None, type=int)

tornado.options.define('timeout_multiplier', default=1.0, type=float)

tornado.options.define('xml_root', default=None, type=str)
tornado.options.define('xml_cache_limit', default=None, type=int)
tornado.options.define('xsl_root', default=None, type=str)
tornado.options.define('xsl_cache_limit', default=None, type=int)
tornado.options.define('xsl_executor_pool_size', default=1, type=int)
tornado.options.define('jinja_template_root', default=None, type=str)
tornado.options.define('jinja_template_cache_limit', default=50, type=int)
tornado.options.define('jinja_streaming_render_timeout_ms', default=50, type=int)

tornado.options.define('sentry_dsn', default=None, type=str, metavar='http://public:secret@example.com/1')

tornado.options.define('max_http_clients', default=100, type=int)
tornado.options.define('max_http_clients_connects', default=None, type=int)
